### PR TITLE
[examples, tasks] feat: add Claude Code recipe for Ascend Triton kernel training

### PR DESCRIPTION
**Pull request by qiangyupei**
## Summary

Add an end-to-end recipe for training coding policies to generate Ascend Triton kernels. Claude Code reads an operator task, implements and verifies a kernel in a per-session sandbox, and iterates on correctness and performance. UniAgent collects the interactions for GRPO training through verl. Claude Code's model requests are served by the training policy through UniAgent's Gateway.

This PR adds a self-contained example without changing shared Framework or Gateway behavior.

## Changes

- Add `uni_agent/tasks/kernel_bench` with the registered `npu_triton_kernel` Task, preprocessing for [DrKernel](https://huggingface.co/datasets/AhNr/dr-kernel-RL), verifier-result collection, and partial-credit rewards.
- Add `examples/claude_code_kernel_task` with task configuration, a thin Task-runner adapter, a remote Docker provider based on UniAgent's Docker sandbox, a trajectory postprocessor, and a synchronous GPU/Megatron training launcher.
- Create and destroy a container per session. Verification acquires NPU leases through host-local file locks; containers on the same host share its device pool, while different hosts use independent pools.
- Include sandbox image build files and an RL adaptation patch for skills and verifier scripts from [CANNBot Skills PR #205](https://gitcode.com/cann/cannbot-skills/pull/205). The build script assembles assets from pinned commit `49451c99df99d0dc31f2a8acb920e46e1ee52105`, with a local-source option for offline builds.
- Support best-snapshot tracking, cooperative early-stop hooks, and assistant-boundary prefix selection through the existing trajectory-postprocessor API. If the best snapshot cannot be reliably mapped to an assistant turn, use the finalized trajectories without best-prefix cropping.
- Support separate validation turn and time budgets, selected through the sample's `metadata.split`.
- Collect workspace snapshots through one bounded, isolated Python call, with an additional call when staged-best recovery is needed. AST, digest, metric-binding, and reward checks remain on the host.

## Usage

See [`examples/claude_code_kernel_task/README.md`](examples/claude_code_kernel_task/README.md) for the complete workflow:

1. Prepare the sandbox image, SSH access, Ascend device/driver mounts, and host-local lock directories.
2. Preprocess the operator dataset into training and validation parquet files.
3. Start the Ray cluster with compatible verl, Megatron, and vLLM dependencies.
4. Run `run_train_gpu.sh` with model/data paths and the remote Docker hosts and NPU device IDs.

The launcher uses existing UniAgent interfaces. Dataset contents, model weights, and sandbox images are not included.

## Compatibility and limitations

- Existing examples and shared Framework/Gateway behavior are unchanged. The only shared Task registration addition is `npu_triton_kernel`; the remote Docker provider is imported by this example's runner.
- Best-prefix attribution depends on readable Claude transcripts and valid snapshot bindings; otherwise selection falls back. Early stopping denies subsequent tool admissions but does not force-kill in-flight requests.
- Privileged NPU containers, sudo-enabled verifier commands, and advisory device locks assume trusted infrastructure.

## Experimental context

A historical development run used Qwen3.6-35B-A3B with 16 tasks × 14 rollouts per training step. Mean training reward increased from 0.3302 at step 1 to 0.4619 at step 31. On a fixed 128-task validation set, reported mean rewards increased from 0.2035 at step 0 to 0.3687 at step 30.

convergence

## Validation

The KernelBench, Task Runner, and Framework CPU test suites passed on the test machine with Python 3.12:

```bash
python -m pytest -q tests/uni_agent/tasks/kernel_bench \
  tests/uni_agent/framework/test_task_runner.py \
  tests/uni_agent/framework/test_generate_sequences_on_cpu.py
# 121 passed, 2 warnings in 25.89s

pre-commit run --all-files --show-diff-on-failure
# All hooks passed.
```

No tests failed or were skipped. The two warnings are Ray deprecation/future warnings. Bash syntax checks also passed for the training launcher and sandbox build script.

## Checklist

- [x] The PR is focused and explains why no separate issue is needed.
- [x] The title follows the repository format and names the owning layers.
- [x] Configuration and deployment steps are documented.
- [x] Compatibility and operational limitations are documented.
- [x] Validate a full training step on the final PR commit.

### CODE REVIEW

File: examples/claude_code_kernel_task/README.md [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+  trajectory_processor.py
+  run_train_gpu.sh           # GPU training, remote NPU verification
+  sandbox/                   # image template, skills and verifier tools
+```
+
+The Task is registered as `npu_triton_kernel`. The runner delegates to the stock `run_task`; the Task uses `uni_agent.agents.claude_code`.
+
+## 1. Prepare the sandbox hosts
+
+The `sandbox/` directory contains the image build files, skills and verifier tools. The base image must already contain Claude Code, Python, Ascend/CANN, torch-npu, Triton Ascend, sudo and the `claude` user. See [sandbox setup](sandbox/README.md); the image must match `task_config_kernel_bench.yaml`.
```
Comment (zackcxb) [inline examples/claude_code_kernel_task/README.md]: sandbox/ is not in the PR

File: examples/claude_code_kernel_task/README.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+```bash
+MODEL_PATH=/models/your-model \
+TRAIN_FILE=/data/triton-agent/train.parquet \
+VAL_FILE=/data/triton-agent/validation.parquet \
+REMOTE_DOCKER_HOSTS=ssh://root@npu-host-01,ssh://root@npu-host-02 \
+EVALUATOR_NPU_DEVICE_IDS=0,1,2,3,4,5,6,7 \
+MAX_PROMPT_LENGTH=24576 \
+MAX_RESPONSE_LENGTH=8192 \
+MAX_MODEL_LEN=32768 \
```
Comment (zackcxb) [inline examples/claude_code_kernel_task/README.md]: Is this default enough for a kernel task? Perhaps provide more details in the experimental context in the PR description?

File: examples/claude_code_kernel_task/README.md [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+bash run_train_gpu.sh
+```
+
+Adjust model parallelism and concurrency to the actual cluster. A 32768-token rollout capacity is a smaller bring-up setting, not a guarantee of sufficient memory.
+
+The launcher submits the repository with Ray `--working-dir`; `RUNTIME_ENV` is optional for deployment-specific dependencies/environment. The installed verl and accelerator packages must match the selected source revision.
+
+- `MAX_RESPONSE_LENGTH` sets rollout `response_length` and Claude's `CLAUDE_CODE_MAX_OUTPUT_TOKENS`. An independent Gateway per-call cap requires the response-length PR.
```
Comment (zackcxb) [inline examples/claude_code_kernel_task/README.md]: Is this PR still needed? The current plan is to leave the configuration to claude code AgentConfig

File: examples/claude_code_kernel_task/README.md [hunk 4]
```diff
@@ -0,0 +1,161 @@
+# Triton Ascend operator-generation recipe
+
+Train an LLM through UniAgent's Claude Code agent, with operator verification on remote Ascend hosts. Use this branch with its pinned verl revision and a matching installed environment; release v0.9.0 alone does not provide the CT APIs required by this baseline. No additional UniAgent patches are required.
```
Comment (Erpim) [inline examples/claude_code_kernel_task/README.md]: Does this content “release v0.9.0 alone does not provide the CT APIs required by this baseline” need to be reflected in the README?

File: uni_agent/tasks/kernel_bench/__init__.py [hunk 5]
```diff
@@ -0,0 +1 @@
+"""KernelBench task, reward, and preprocessing support."""
```
Comment (Erpim) [inline uni_agent/tasks/kernel_bench/__init__.py:1]: Are the training set and validation set two separate tasks? Should they be implemented separately?

File: uni_agent/tasks/kernel_bench/preprocess.py [hunk 6]
```diff
@@ -0,0 +1,1064 @@
...
+import re
+import tokenize
+import uuid
+from collections.abc import Mapping
+from dataclasses import dataclass, field
+from pathlib import Path, PurePosixPath
+from typing import Any
+
+SYSTEM_PROMPT = "You are a coding agent optimizing Triton Ascend operators in an isolated NPU sandbox."
+USER_PROMPT = """Implement the Ascend Triton operator in {workspace}/src/{op_name}.py.
```
Comment (Erpim) [inline uni_agent/tasks/kernel_bench/preprocess.py:17]: Does USER_PROMPT support customization for different agents?

File: examples/claude_code_kernel_task/README.md (hunk 1 above)
Comment (qiangyupei) [reply examples/claude_code_kernel_task/README.md]: I will add it soon

File: examples/claude_code_kernel_task/README.md (hunk 2 above)
Comment (qiangyupei) [reply examples/claude_code_kernel_task/README.md]: The doc has been updated

File: examples/claude_code_kernel_task/README.md (hunk 3 above)
Comment (qiangyupei) [reply examples/claude_code_kernel_task/README.md]: The doc has been updated

File: examples/claude_code_kernel_task/README.md (hunk 4 above)
Comment (qiangyupei) [reply examples/claude_code_kernel_task/README.md]: The doc has been updated

File: uni_agent/tasks/kernel_bench/__init__.py (hunk 5 above)
Comment (qiangyupei) [reply uni_agent/tasks/kernel_bench/__init__.py:1]: The training and validation sets come from the same dataset, which provides both train and validation Parquet files. The only differences between the training and validation configurations are max_turn and the time budget. I don't think separate handling is necessary.

File: examples/claude_code_kernel_task/run_train_gpu.sh [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+  actor_rollout_ref.rollout.log_prob_max_token_len_per_gpu=${infer_ppo_max_token_len}
+  actor_rollout_ref.rollout.log_prob_use_dynamic_bsz=${use_dynamic_bsz}
+  actor_rollout_ref.rollout.multi_turn.enable=True
+  actor_rollout_ref.rollout.multi_turn.max_parallel_calls=1
+  ++actor_rollout_ref.rollout.multi_turn.format=${TOOL_PARSER}
+  actor_rollout_ref.rollout.agent.num_workers=${AGENT_WORKERS}
+  ++actor_rollout_ref.rollout.agent.agent_loop_manager_class=uni_agent.framework.entry.AgentFrameworkRolloutAdapter
+  ++actor_rollout_ref.rollout.custom.agent_framework.gateway_count=${GATEWAY_COUNT}
+  "++actor_rollout_ref.rollout.custom.agent_framework.log_dir=${AGENT_LOG_DIR}"
+  ++actor_rollout_ref.rollout.custom.agent_framework.use_reward_loop_worker=False
```
Comment (zackcxb) [inline examples/claude_code_kernel_task/run_train_gpu.sh]: this field is not needed anymore

File: examples/claude_code_kernel_task/sandbox/template/CLAUDE.md [hunk 8]
```diff
@@ -0,0 +1,74 @@
+# Triton Ascend KernelBench Rollout
```
Comment (yyDing1) [inline examples/claude_code_kernel_task/sandbox/template/CLAUDE.md:1]: Is this folder passed to the Claude Code Agent to run kernel tasks?
We'd better not maintain this in the repo, consider compressing it into a tarball and uploading it to a public website like Hugging Face?

File: examples/claude_code_kernel_task/runner.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+        lock_timeout=evaluator_npu_lock_timeout,
+    )
+    task_config = copied.get("task")
+    if not isinstance(task_config, dict):
+        raise ValueError("run_triton_task requires tools_kwargs['task']")
+    if task_config.get("metadata", {}).get("split") == "validation":
+        task_config.setdefault("agent", {}).update(max_turns=validation_max_turns, run_timeout=validation_run_timeout)
+    if max_response_length is not None:
+        extra_env = task_config.setdefault("agent", {}).setdefault("extra_env", {})
+        extra_env["CLAUDE_CODE_MAX_OUTPUT_TOKENS"] = str(max_response_length)
```
Comment (zackcxb) [inline examples/claude_code_kernel_task/runner.py]: Why not use extra_env directly?

File: uni_agent/tasks/kernel_bench/preprocess.py (hunk 6 above)
Comment (zackcxb) [reply uni_agent/tasks/kernel_bench/preprocess.py:17]: We can override them in task_config.prompt_template

File: uni_agent/tasks/kernel_bench/reward.py [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+
+def attach_reward(metrics: dict[str, Any], weights: dict[str, float] | None = None) -> dict[str, Any]:
+    result = dict(metrics)
+    components = reward_breakdown(result, weights)
+    result["reward"] = components["total"]
+    result["reward_components"] = components
+    return result
+
+
+def compute_score(data_source: str, solution_str: str, ground_truth: Any, extra_info=None) -> dict[str, float]:
```
Comment (zackcxb) [inline uni_agent/tasks/kernel_bench/reward.py]: Is this function still used? It doesn't seem quite right. Consider use uni_agent.framework.task_runner.score_from_runner_result

File: examples/claude_code_kernel_task/runner.py (hunk 9 above)
Comment (qiangyupei) [reply examples/claude_code_kernel_task/runner.py]: because the user may not set the value of max_response_length

File: uni_agent/tasks/kernel_bench/reward.py (hunk 10 above)
Comment (qiangyupei) [reply uni_agent/tasks/kernel_bench/reward.py]: The function has been removed

File: examples/claude_code_kernel_task/runner.py (hunk 9 above)
Comment (qiangyupei) [reply examples/claude_code_kernel_task/runner.py]: The CLAUDE_CODE_MAX_OUTPUT_TOKENS env variable setting has been moved to extra_env

File: examples/claude_code_kernel_task/sandbox/template/CLAUDE.md (hunk 8 above)
Comment (qiangyupei) [reply examples/claude_code_kernel_task/sandbox/template/CLAUDE.md:1]: Yes. The template is copied into sandbox workspace, where Claude Code reads the task instructions and skills to implement and verify kernels.
To avoid vendoring upstream assets, I’ve updated this PR to fetch skills and verifier scripts from a pinned CANNBot commit (https://gitcode.com/cann/cannbot-skills/pull/205) and only apply a local RL adaptation patch at image-build time. Only recipe-specific instructions (i.e., CLUADE.md and INSTRUCTION.md, small files) and wrapper scripts (e.g., sandbox/tools) remain directly maintained here. Would this address your concern?
